tower-http/src/catch_panic.rs

```diff
@@ -0,0 +1,202 @@
+//! TODO(david): docs
+
+use bytes::Bytes;
+use futures_core::ready;
+use futures_util::future::{CatchUnwind, FutureExt};
+use http::{Request, Response, StatusCode};
+use http_body::{combinators::UnsyncBoxBody, Body, Full};
+use pin_project_lite::pin_project;
+use std::{
+    any::Any,
+    borrow::Cow,
+    future::Future,
+    panic::AssertUnwindSafe,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies the [`CatchPanic`] middleware that catches panics and converts them into
+/// `500 Internal Server` responses.
+///
+/// See the [module docs](self) for an example.
+#[derive(Debug, Clone, Copy, Default)]
+#[non_exhaustive]
+pub struct CatchPanicLayer;
+
+impl CatchPanicLayer {
+    /// Create a new `CatchPanicLayer`.
+    pub fn new() -> Self {
+        CatchPanicLayer {}
+    }
+}
+
+impl<S> Layer<S> for CatchPanicLayer {
+    type Service = CatchPanic<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CatchPanic::new(inner)
+    }
+}
+
+/// Middleware that catches panics and converts them into `500 Internal Server` responses.
+///
+/// See the [module docs](self) for an example.
+#[derive(Debug, Clone, Copy)]
+pub struct CatchPanic<S> {
+    inner: S,
+}
+
+impl<S> CatchPanic<S> {
+    /// Create a new `CatchPanic`.
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+
+    define_inner_service_accessors!();
+
+    /// Returns a new [`Layer`] that wraps services with a `CatchPanic` middleware.
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer() -> CatchPanicLayer {
+        CatchPanicLayer::new()
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CatchPanic<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Response = Response<UnsyncBoxBody<Bytes, ResBody::Error>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
```

Comment:
Not sure, but should this also catch panics? 🤔

Reply:
Hm so thats a good question actually. I'm actually thinking no, panics from `poll_ready` shouldn't be handled for a few reasons:

- We cannot return a response from `poll_ready` so we'd have to return an error. That would require changing the error type to `BoxError` which means a service that previously returned `Infallible` would now return `BoxError`. [This makes it more annoying to use with axum.](https://docs.rs/axum/latest/axum/error_handling/index.html)
- If `poll_ready` returns an error the `Service` contract dictates that the service must be discarded and recreated. Hyper handles that and does the same thing regardless if `poll_ready` errors or panics. So since we cannot return a response it _kinda_ doesn't matter what we do 🤷 
- In practice its rare for services to actually care about `poll_ready` so panics from here should be very rare. I would guess they're much more common from `call`.

@hawkw What do you think?

Reply:
Yeah, I don't think we should catch panics here. Those reasons track for me.